nixos/modules/services/search/quickwit.nix

```diff
@@ -0,0 +1,91 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.quickwit;
+  settingsFormat = pkgs.formats.yaml { };
+  configFile = settingsFormat.generate "quickwit.yaml" cfg.settings;
+in
+{
+  options.services.quickwit = {
+    package = mkOption {
+      type = types.package;
+      description = "Quickwit package to use.";
+      default = pkgs.quickwit;
+    };
+
+    enable = mkEnableOption "Quickwit search engine";
+
+    environmentFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = "Environment file to pass to the systemd service.";
+    };
+
+    settings = mkOption {
+      description = "Quickwit node configuration, see <https://quickwit.io/docs/configuration/node-config>";
+      default = { };
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+
+        options = {
+
+          version = mkOption {
+            type = types.str;
+            default = "0.7";
+            description = "Config file version.";
+          };
+
+          listen_address = mkOption {
+            type = types.str;
+            default = "127.0.0.1";
+            description = "The IP address or hostname that Quickwit service binds to for starting REST and GRPC server and connecting this node to other nodes.";
+          };
+
+          rest = {
+            listen_port = mkOption {
+              type = types.port;
+              default = 7280;
+              description = "The port on which the REST API listens for HTTP traffic.";
+            };
+          };
+
+          data_dir = mkOption {
+            type = types.path;
+            default = "/var/lib/quickwit";
+            description = "Path to directory where data (tmp data, splits kept for caching purpose) is persisted. This is mostly used in indexing.";
+          };
+
+        };
+      };
+    };
+  };
+  config = mkIf cfg.enable {
+    users.users.quickwit = {
+      description = "Quickwit daemon user";
+      isSystemUser = true;
+      group = "quickwit";
+    };
+    users.groups.quickwit = {};
+    system.activationScripts.makeQuickwitDir = ''
```

Comment:
I suspect that the current way to do that is with `systemd.tmpfiles`, see e.g.

https://github.com/NixOS/nixpkgs/blob/0638fe2715d998fa81d173aad264eb671ce2ebc1/nixos/modules/services/monitoring/munin.nix#L402-L407

But do it a bit better than done there: Refer to `users.users.quickwit.name` instead of hardcoding the user name in multiple places.

Reply:
Hum this is weird to me though it's called tmpfiles. 
https://search.nixos.org/options?channel=23.11&show=systemd.tmpfiles.rules&from=0&size=50&sort=relevance&type=packages&query=systemd.tmpfiles.rules it's says it's for volatile dir and data dir is everything but volatile.

Reply:
I did like that because I saw it in another package (I don't exactly remember which one though).

Reply:
> Hum this is weird to me though it's called tmpfiles.

I find that weird, too.

[Systemd docs](https://www.freedesktop.org/software/systemd/man/latest/tmpfiles.d.html) say this about it:

> `StateDirectory=`, `CacheDirectory=`, `LogsDirectory=`, and `ConfigurationDirectory=` should be used to create directories under `/var/lib/`, `/var/cache/`, `/var/log/`, and `/etc/`.
> `tmpfiles.d` should be used for files whose lifetime is independent of any service or requires more complicated configuration.

So indeed it seems like `StateDirectory=` would be better than both `tmpfiles` and `activationScripts`.